urlfb310a588a_alokhan_LinkStateNetworkSimulation_tgz-pJ8-reso_examples_static_routing_DemoJ8/src/reso/utilities/NetworkGrapher.java

```diff
@@ -0,0 +1,185 @@
+	public static final double EDGE_LENGTH= 2; // default=1
+	
+	private static int getMetric(Node node, HardwareInterface<?> iface) {
+		if (!(node instanceof IPHost))
+			return Integer.MAX_VALUE;
+		IPHost ip_node= (IPHost) node;
+		IPInterfaceAdapter ip_iface= ip_node.getIPLayer().getInterfaceByName(iface.getName());
+		return ip_iface.getMetric();
+	}
+	
+	/* This method generates a text file representation of the network graph
+	 * in AT&T's dot language (see http://www.graphviz.org).
+	 * 
+	 * The text file can then be converted a graph image in .png, .eps or .pdf format
+	 * using one of the graphviz tools: "neato" and "fdp" are the preferred ones.
+	 */
+	public static void toGraphviz(Network network, PrintWriter pw) {
+		pw.println("graph \"reso\" {");
+		pw.print("  graph [");
+		pw.print("outputorder=nodesfirst");
+		pw.print(", overlap=\"scale\"");
+		pw.print(", mindist=" + 5);
+		pw.print(", K=" + 1);
+		pw.println("] ;");
+		pw.print("  edge [");
+		pw.print("labelfontsize=" + EDGE_LABEL_FONT_SIZE);
+		pw.print(", fontsize=" + 8);
+		pw.print(", labeldistance=" + EDGE_LABEL_DISTANCE);
+		//pw.print(", labelangle=" + 30);
+		pw.print(", len=" + EDGE_LENGTH);
+		pw.println("] ;");
+		pw.print("  node [");
+		pw.print("fontsize=" + NODE_LABEL_FONT_SIZE);
+		pw.println("] ;");
+		
+		for (Node n: network.getNodes()) {
+			pw.print("  \"" + n.name + "\" [");
+			pw.print("shape=" + NODE_SHAPE);
+			pw.println("] ;");
+		}
+		
+		for (Node n: network.getNodes()) {
+			for (HardwareInterface<?> iface: n.getInterfaces()) {
+				Link<?> l= iface.getLink();
+				HardwareInterface<?> hi= l.getHead(); 
+				Node hn= hi.getNode();
+				HardwareInterface<?> ti= l.getTail();
+				Node tn= ti.getNode();
+
+				if (tn == n)
+					continue;
+				
+				if (hn.name.compareTo(tn.name) > 0) {
+					HardwareInterface<? extends Message> tmpi= hi;
+					hi= ti;
+					ti= tmpi;
+					Node tmpn= hn;
+					hn= tn;
+					tn= tmpn;
+				}
+				
+				int hmetric= getMetric(hn, hi);
+				int tmetric= getMetric(tn, ti);
+					
+				// Note: with undirected edges, graphviz will consider the left node as the tail and
+				// the right node as the head. It's a bit confusing... but that's the convention we need
+				// to follow
+				pw.print("  \"" + tn.name + "\" -- \"" + hn.name + "\" [");
+				pw.print("headlabel=\"" + hi.getName() + (hi.isActive()?"":" (down)") + "\"");
+				pw.print(", taillabel=\"" + ti.getName() + (ti.isActive()?"":" (down)") + "\"");
+				pw.print(", label=\"(" + ((hmetric == Integer.MAX_VALUE)?"?":hmetric) + "/" + 
+						((tmetric == Integer.MAX_VALUE)?"?":tmetric)+ ")\"");
+				pw.println("] ;");
+				
+			}
+		}
+		pw.println("}");
+	}
+	
+	private static IPRouteEntry getRoute(Node n, IPAddress dst) {
+		if (!(n instanceof IPHost))
+			return null;
+		return ((IPHost) n).getIPLayer().getRouteTo(dst);
+	}
+	
+	public static void toGraphviz2(Network network, IPAddress dst, PrintWriter pw) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 125-132 into a separate method.